File: node/tests/examples/CounterAggregate.test.ts
import { describe, it, expect } from 'vitest'
import { CounterAggregate } from '../../src/examples/CounterAggregate'

describe('CounterAggregate', () => {
  it('should replay from history', () => {
    const aggregateId = "counter-1"
    const counter = new CounterAggregate(aggregateId)
    counter.increment()

    const replayed = new CounterAggregate(aggregateId)
    replayed.replay(counter.uncommittedEvents)

    expect(replayed.value).toBe(counter.value)
  })
})

describe('CounterAggregate', () => {
  it('should increment', () => {
    const aggregateId = "counter-1"
    const counter = new CounterAggregate(aggregateId)

    counter.increment()

    expect(counter.value).toBe(1)
  })
})

describe('CounterAggregate', () => {
  it('should decrement', () => {
    const aggregateId = "counter-1"
    const counter = new CounterAggregate(aggregateId)

    counter.decrement()

    expect(counter.value).toBe(-1)
  })
})

File: cpp/tests/CounterAggregate.test.cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch_test_macros.hpp>

#include "../src/core/Event.hpp"
#include "../src/core/EventStore.hpp"
#include "../src/examples/CounterAggregate.hpp"

// Use the simple CounterAggregate implementation from examples
// (A minimal test that focuses on replay and append)

TEST_CASE("CounterAggregate increments and produces events") {
    CounterAggregate agg("counter-1");
    agg.increment();
    agg.increment();
    agg.decrement();

    REQUIRE(agg.state().value == 1);
    REQUIRE(agg.uncommittedEvents().size() == 3);
}

TEST_CASE("InMemoryEventStore append/load and replay") {
    InMemoryEventStore<std::variant<CounterIncremented, CounterDecremented>> store;
    CounterAggregate agg("counter-2");
    agg.increment();
    agg.increment();

    auto events = agg.uncommittedEvents();
    store.appendEvents(agg.id(), events, 0);

    auto history = store.loadEvents(agg.id());

    CounterAggregate loaded("counter-2");
    loaded.replay(history);

    REQUIRE(loaded.state().value == 2);
}
